### PR TITLE
doc: Add a tutorial how to write plugins

### LINKED ISSUE #287: Create documentation regarding plugins
DNF5 supports creating two types of plugins:
- DNF5 plugins
  - **active** ones implementing one or more commands
- LIBDNF plugins
  - **passive** ones implementing additional logic and hooks to DNF5

So far we don't have documentation on how to implement and deploy new plugins.

### PR DESCRIPTION
**Pull request by jan-kolarik**
Documentation updates cover the following aspects:

- DNF5 and LIBDNF5 plugins
- Example template codes
- Purpose of plugins
- Deployment description

Closes #287.

### CODE REVIEW

File: doc/tutorial/plugins/dnf5-plugins.rst [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+.. note::
+    During the startup of the dnf5 application, the plugin library scans
+    the files in the configured plugins directory to check for the presence
+    of any plugins implementing the common ``dnf5::IPlugin`` interface.
+    If such plugins are found, they are loaded into memory along with all
+    the commands they implement.
+
+Each plugin source is structured in its own directory within the
+``dnf5-plugins`` folder. See other plugins, such as ``builddep_plugin``,
+to see how the result should look.
```
Comment (inknos) [inline doc/tutorial/plugins/dnf5-plugins.rst:47]: I would rather put something like:

```
$ tree dnf5-plugins/builddep_plugin/
dnf5-plugins/builddep_plugin/
├── builddep_cmd_plugin.cpp
├── builddep.cpp
├── builddep.hpp
└── CMakeLists.txt
```

What do you think?

File: doc/tutorial/plugins/dnf5-plugins.rst [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+Debugging Tips
+--------------
+
+To test your freshly built plugin, redirect dnf5 to load it by setting
+the ``DNF5_PLUGINS_DIR`` environmental variable to your build directory,
+e.g., ``DNF5_PLUGINS_DIR=/home/user/dnf5/build/dnf5-plugins/template_plugin``.
+
+Ensure effective debugging by building the project with debugging symbols
+(``-DCMAKE_BUILD_TYPE=Debug``).
```
Comment (inknos) [inline doc/tutorial/plugins/dnf5-plugins.rst:87]: I love this section. I wish we had more tips at the end of each section to enrich our documentation.

File: doc/tutorial/plugins/dnf5-plugins.rst [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+``Provides: dnf5-command(template)``.
+
+You should also include a documentation man page describing the usage of your
+plugin in detail. There are several places which need to be interconnected:
+
+* ``doc/dnf5_plugins/template.8.rst``: Add a new man page for your plugin with the respective name.
+* ``doc/dnf5_plugins/index.rst``: Reference the new plugin in the dnf5 plugins page.
+* ``doc/dnf5.8.rst``: Reference the new plugin in the main dnf5 man page.
+* ``doc/CMakeLists.txt`` and ``doc/conf.py.in``: Integrate with Sphinx documentation.
+* ``dnf5.spec``: Include the new man page in the ``dnf5-plugins`` package.
```
Comment (inknos) [inline doc/tutorial/plugins/dnf5-plugins.rst:77]: if this is a checklist I would say it in the sentence right before. something simple like.

"be sure to complete all these steps"

File: doc/tutorial/plugins/libdnf5-plugins.rst [hunk 4]
```diff
@@ -0,0 +1,128 @@
+.. _libdnf5 plugins tutorial:
+
+LIBDNF5 Plugins
+===============
+
+These plugins enable changes to the existing dnf5 workflow.
+Modify the library's behavior at defined points, implement new logic,
+or trigger the loading of additional plugins by implementing prepared
+callbacks.
```
Comment (inknos) [inline doc/tutorial/plugins/libdnf5-plugins.rst:9]: What do you think of a rephrase like this?

"Possible uses of DNF5 passive plugins are modifying the libdnf5's behavior at specific breakpoints, changing or implementing the logic, or triggering the loading of additional plugins via prepared callbacks."

File: doc/tutorial/plugins/dnf5-plugins.rst [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+The core commands in ``dnf5/commands`` implement essential package manager
+functionality, primarily focused on installing and managing packages. They are
+integral to the dnf5 package.
+
+Plugin commands implement additional features that extend the capabilities
+of dnf5. They are packaged separately for optional installation by users who
+need these additional features.
+
+Writing the Plugin
```
Comment (inknos) [inline doc/tutorial/plugins/dnf5-plugins.rst:23]: What would you think about changing the header to "Writing an Active Plugin"?

File: doc/tutorial/plugins/libdnf5-plugins.rst [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+Writing the Plugin
```
Comment (inknos) [inline doc/tutorial/plugins/libdnf5-plugins.rst:11]: Same for this: "Writing a Passive Plugin". WDYT?

File: doc/tutorial/plugins/libdnf5-plugins.rst [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+------------------
+
+Similar to the :ref:`dnf5 plugins tutorial`, we have to implement
```
Comment (inknos) [inline doc/tutorial/plugins/libdnf5-plugins.rst:14]: s/Similar/Similarly

File: doc/tutorial/plugins/libdnf5-plugins.rst [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+the ``libdnf5::plugin::IPlugin`` interface and override the hooks
+where additional logic is inserted within the dnf5 application lifecycle.
```
Comment (inknos) [inline doc/tutorial/plugins/libdnf5-plugins.rst:16]: It is a bit convoluted in the end. I would like to simplify it like:

"Similarly to the plugins tutorial, we have to implement the `libdnf5::plugin::IPlugin` interface and override the hooks to alter DNF5's logic."

File: doc/tutorial/plugins/libdnf5-plugins.rst [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+In the following code block, a simple example plugin adds logic at two
+different places: after the ``libdnf5::Base`` object is prepared
+and before the actual transaction starts:
```
Comment (inknos) [inline doc/tutorial/plugins/libdnf5-plugins.rst:20]: Also here. I would maybe split the sentence to make it more clear.

"In the following code block, a simple example plugin adds logic in two
different steps.
The first step is after preparing the ``libdnf5::Base`` object.
The second step is before the start of the transaction."

File: doc/tutorial/plugins/libdnf5-plugins.rst [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+.. literalinclude:: ../../templates/libdnf5-plugin/template.cpp
+    :language: c++
+    :linenos:
+
+Each plugin is structured in its own directory within the ``libdnf5-plugins``
+folder. Review other plugins, such as ``actions``, to understand the expected
+structure.
```
Comment (inknos) [inline doc/tutorial/plugins/libdnf5-plugins.rst:28]: same as before

WDYT of this?
```
$ tree libdnf5-plugins/actions/
libdnf5-plugins/actions/
├── actions.conf
├── actions.cpp
└── CMakeLists.txt
```

File: doc/tutorial/plugins/libdnf5-plugins.rst [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+configuration by setting the ``LIBDNF_PLUGINS_CONFIG_DIR`` environmental
+variable to your directory with the plugin's configuration, e.g.,
+``LIBDNF_PLUGINS_CONFIG_DIR=/home/user/dnf5/libdnf5-plugins/template.``
+This can also be overridden by the ``pluginconfpath`` configuration option.
+
+Next, configure the directory where the plugin binaries are located using
+the ``pluginpath`` configuration option.
+
+For effective interception of your plugin code at defined breakpoints, ensure
+you are building the project with debugging symbols (``-DCMAKE_BUILD_TYPE=Debug``)."
```
Comment (inknos) [inline doc/tutorial/plugins/libdnf5-plugins.rst:128]: can these tips be a single tip section to deduplicate info in the documentation?

**Review by inknos — COMMENTED** [review]
I only have a small review on the code. Then LGTM

File: doc/templates/dnf5-plugin/template_cmd_plugin.cpp [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace {
+
+/* Enter the name of your new plugin. */
+constexpr const char * PLUGIN_NAME{"template"};
+
+/* Fill in the version of your plugin. */
+constexpr PluginVersion PLUGIN_VERSION{.major = 1, .minor = 0, .micro = 0};
+
+/* Add information about yourself and a description of the plugin. */
```
Comment (j-mracek) [inline doc/templates/dnf5-plugin/template_cmd_plugin.cpp:15]: Is it mandatory or optional?

File: doc/templates/dnf5-plugin/template_cmd_plugin.cpp [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+#include <dnf5/iplugin.hpp>
+
+using namespace dnf5;
+
+namespace {
+
+/* Enter the name of your new plugin. */
+constexpr const char * PLUGIN_NAME{"template"};
+
+/* Fill in the version of your plugin. */
```
Comment (j-mracek) [inline doc/templates/dnf5-plugin/template_cmd_plugin.cpp:12]: Would be possible to write why this is important, how it is used or whether is it mandatory?

File: doc/templates/libdnf5-plugin/template.conf [hunk 14]
```diff
@@ -0,0 +1,7 @@
+[main]
+name = template
+enabled = yes
+version = 1.0.1
+
+[custom]
+website = https://my-hosting/template-plugin.org
```
Comment (j-mracek) [inline doc/templates/libdnf5-plugin/template.conf:7]: I was unable to find any description to the section and a key `website`

File: doc/templates/libdnf5-plugin/template.cpp [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+/* Fill in the version of your plugin. */
+constexpr plugin::Version PLUGIN_VERSION{.major = 1, .minor = 0, .micro = 0};
+
+/* Add information about yourself and a description of the plugin. */
+constexpr const char * attrs[]{"author.name", "author.email", "description", nullptr};
+constexpr const char * attrs_value[]{"Fatima Freedom", "dummy@email.com", "Plugin description."};
+
+
+class TemplatePlugin : public plugin::IPlugin {
```
Comment (j-mracek) [inline doc/templates/libdnf5-plugin/template.cpp:21]: I would suggest to add a comment for the class and each of its methods, including inheritance.

File: doc/templates/libdnf5-plugin/template.cpp [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    transaction.set_description("This is our important transaction.");
+}
+
+}  // namespace
+
+PluginAPIVersion libdnf_plugin_get_api_version(void) {
+    return PLUGIN_API_VERSION;
+}
+
+const char * libdnf_plugin_get_name(void) {
```
Comment (j-mracek) [inline doc/templates/libdnf5-plugin/template.cpp:73]: Is it mandatory?

File: doc/templates/dnf5-plugin/template_cmd_plugin.cpp [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+constexpr PluginVersion PLUGIN_VERSION{.major = 1, .minor = 0, .micro = 0};
+
+/* Add information about yourself and a description of the plugin. */
+constexpr const char * attrs[]{"author.name", "author.email", "description", nullptr};
+constexpr const char * attrs_value[]{"Fred Fedora", "dummy@email.com", "Plugin description."};
+
+class TemplateCmdPlugin : public IPlugin {
```
[the remark this answers is not in the document]
Comment (jan-kolarik) [reply doc/templates/dnf5-plugin/template_cmd_plugin.cpp:16]: I've provided some additional brief docstrings for the `IPlugin` classes in dnf5 and libdnf5 namespaces. Please let me know if more detailed information would be appropriate.

**Comment by jan-kolarik** [comment]
I've tried to address all comments. Please let me know if I've missed anything.

**Comment by j-mracek** [comment]
LGTM

**Pull request merged**